Add command to disable notations

herbelin opened the pull request:
**Kind:** feature

This extends #703 on selective deactivation of notations by adding:
- concrete syntax for deactivation/reactivation
- registration in vo files
- support for deactivation of abbreviations

Added: Now depends on #12690 (merged) and #12950 (merged) and supports selective only-parsing or only-printing deactivation.

#### Concrete syntax for deactivation/reactivation

The currently proposed syntax is:
- for notations
  ```
  Deactivate Notation "_ + _" (in custom foo) : scope.
  Activate Notation "_ + _" (in custom foo) : scope.
  ```
  where both `(in custom foo)` and `: scope` are optional, defaulting to `constr` and lonely notation respectively
- for abbreviations
  ```
  Deactivate Notation foo.
  Activate Notation foo.
  ```

I wonder whether this could not be reusing `Set`/`Unset` even if a non standard use of it.

#### Adding support for deactivation of abbreviations

We obtain this by adding a new `remove` function to the name table.

#### Miscellaneous

In passing, we include the `only printing` notations in `Print Scopes`.

Fixes #4803

#### Status

- [x] I did not port nor give syntax to `show_inactive` yet. Not clear what would be the best user interface for it (commands like `Print Scope` or `Locate` show the enabled/disabled status of notations).
- [x] I did not do anything about `deactivate_scope`/`reactivate_scope`/`show_scope` (no syntax, no registration in vo) (now handled by `Enable Notation; Disable Notation` provided only a scope)
- [x] More testing needed (test-suite updated)

- [X] Added / updated test-suite
- [x] Corresponding documentation was added / updated
- [x] Entry added in the changelog

### Overlay (to be merged in sync with the PR):

* https://github.com/ejgallego/coq-serapi/pull/289

Zimmi48 commented:
> I wonder whether this could not be reusing `Set`/`Unset` even if a non standard use of it.

I'm not very fond of this idea. On the contrary, I'd like to move the last non-standard use of `Set` (`Warnings`) to a new mechanism...

proux01 commented:
Is there a way to activate/deactivate only for parsing/printing? would there be a use for that? (if that makes any sense)

herbelin commented:
> Is there a way to activate/deactivate only for parsing/printing? would there be a use for that? (if that makes any sense)

Funny, that's what I was precisely thinking too. That would be easy to do too. Syntax could be e.g.:
```
Deactivate Notation "_ + _" (in custom foo, only parsing) : scope.
Deactivate Notation "_ + _" (in custom foo, only printing) : scope.
```

I don't like so much `Deactivate` which is a new command name to remember, but I don't have a good alternative.

Zimmi48 commented:
> I don't like so much `Deactivate` which is a new command name to remember, but I don't have a good alternative.

It looks to me like the `Deactivate` syntax is very close to the `Notation` syntax. Then, an option is to reuse the `Notation` command but with an attribute. BTW, I think "disable" is a simpler word than "deactivate" that conveys the same meaning. So this would give:

```
#[ disable ] Notation "_ + _" (in custom foo, only parsing) : scope.
#[ disable ] Notation foo.
#[ disable ] Infix "/\".
```

BTW, some other command with a very close syntax is `Reserved Notation`. It could also become an attribute:

```
#[ reserve ] Notation "x \/ y" (at level 85, right associativity).
#[ reserve ] Infix  "∘" (at level 40, left associativity).
```

herbelin commented:
Following @proux' remark, I propose to enforce the following model for activation/deactivation:
- To any triple of a constr or custom entry, notation string and scope can be associated at most one "main" notation + an arbitrary number of only-printing. This actually reflects the current situation which indeed tolerates for convenience an arbitrary number of distinct interpretation for only-printing notations
- A main notation consists of either an only-parsing notation or of a "regular" notation with both parsing and printing.

When there is only one notation, whether it is its parsing rule which has to deactivated (if any), its printing rule (if any), or both, it would be enough to give the string and, if not a lonely notation, the scope, and, if in an entry different from constr, the name of the custom entry. This would mean (as given above) a command of the form (whatever the exact syntax):
```
Deactivate Notation "_ + _" (in custom foo, only printing) : scope.
Deactivate Notation "_ + _" (in custom foo, only parsing) : scope.
```
with all of `in custom foo`, or `only printing`/`only parsing`, or `: scope` optional.

When there are several, we would need however to specify the interpretation as in:
```
Deactivate Notation "_ + _" := term (in custom foo, only printing) : scope.
```
Should we tolerate the syntax without an interpretation and make it default to the main interpretation when the only printing flag is not used (since there would then be only one not-only-printing notation)?

silene commented:
I do not intuitively understand what `Deactivate Notation "_ + _" (in custom foo, only printing) : scope` means. Does it mean that the notation is now only used for parsing? Or that it is used only used for printing?

Thus, associating some change of behavior to a negative command (i.e., `Deactivate`) seems like a bad idea. It would be better to make it part of a positive command (e.g., `Activate`).

gares commented:
While I clearly see the interest of this feature but I find the user interface a bit cumbersome.

If I'm a regular user how am I supposed to guess all this data: `Deactivate Notation "_ + _" (in custom foo, only printing) : scope`? I mean, I may want to see what is behind "_ + _" in the current goal, but asking me more than that seems a bit too much.

If I'm a library designer I would know how to find that metadata, but being forced to repeat all that metadata seems painful (copy pasting it may double maintenance cost). I'd be happier to give (the few) notations that I want to disable a name or a category and mention that name/category in another file were I gather, say, the settings for beginners.

herbelin commented:
> I do not intuitively understand what `Deactivate Notation "_ + _"` (in custom foo, only printing) : scope means. Does it mean that the notation is now only used for parsing? Or that it is used only used for printing?

The idea was that it stopped to be used for printing. I agree that this combination of a negative command and "only" is unclear.

> If I'm a regular user how am I supposed to guess all this data: `Deactivate Notation "_ + _" (in custom foo, only printing) : scope`? 

Use of custom entry remains seldom, so, in the general case, if you want to deactivate both printing and parsing, it would just be something like `Deactivate Notation "_ + _" : nat_scope`.

> I mean, I may want to see what is behind `_ + _` in the current goal, but asking me more than the characters I see seems too much.

That's a very legitimate use but I don't see how to do it easily, if not with the help of the UI. Especially, if it is in a complex expression, the scope is generally implicit and difficult to reconstruct.

For your usage, I would rather rely on the UI, either by sending it a rich output with hidden annotation revealed when hovering over the notation with the mouse, or the UI sending a request to Coq when hovering on a given location of the goal. I understood that this was something @maximedenes was working in the context of LSP. 

From a textual interface, I don't see how to avoid, say, some `Locate "_ + _"` followed by the appropriate `Deactivate`, because I don't see how to magically know which `+` of which grammar and of which scope we are talking about when we see it on screen, especially if we are an ousider and not familiar with the given context.

> I'd be happier to give (the few) notations that I want to disable a name or a category and mention that name/category in another file were I gather, say, the settings for beginners.

Could you elaborate on this? What kind of user API would this correspond to?

More generally, let me remind that I'm ok with different kinds of concrete syntax, but we have already to clarify the level of fine-tuning and simplicity we need (parsing or printing or both, one "+" in some scope and entry, or all "+" of all scopes and entry, ...).

Actually, maybe should we try and experiment and possibly revise afterwards.

gares commented:
> That's a very legitimate use but I don't see how to do it easily, if not with the help of the UI. Especially, if it is in a complex expression, the scope is generally implicit and difficult to reconstruct.

That is exactly my point. The non-expert user may just say `Disable "+"` and have all the notations having `+` as a token to be disabled. This is the user API I was hinting at. It may be too rough for an expert user, but a beginner can surely use it, while the one proposed in the PR seems too hard to me.

> Could you elaborate on this? What kind of user API would this correspond to?

My use case is that in Math Comp we have too many notations for a beginner even in simple files like ssrnat.
When we teach, it helsp to disable some of them, and introduce them "slowly".

For example last time we had a notation `= true` to display the `is_true` coercion during the first session; we disabled `x < y := S x <= y` and `a < b < c := a < b && b < c` and more of the like during the second one. 

All this requires to fork Coq, fork MathComp, build a custom jscoq... not very easy. If I could sneak an attribute next to the notation for, say, `<` (and all the others that I find disturbing for a beginner) and disable them in one go from a `beginner.v` file, I could just ask the students to load that file!

herbelin commented:
> The non-expert user may just say `Disable "+"

Yes, we can provide that also.

> If I could sneak an attribute next to the notation for, say, `<` (and all the others that I find disturbing for a beginner) and disable them in one go from a `beginner.v` file, I could just ask the students to load that file!

Rather than anticipating for each notation how it would have to behave in a `beginner.v` file, wouldn't be as easier for a teacher, and also less intrusive in the original development, and also more flexible, that the teacher aggregates a set of "surgical" variants of `Disable` in this `beginner.v` file?

jfehrle reviewed (changes requested):
When you're ready, would you add a changelog entry and documentation?  You will probably need to add an ".. insertprodn / .. prodn" in the rst for some of the new productions.  Ask me if you have questions.

Then run `make doc_gram_rsts` and add its changes to the PR.  The tool is described in `doc/tools/docgram/README.md`

herbelin commented:
Finally rebased on top of its dependencies! I'm still a bit in the fog about the user API and syntax though.

1. `Deactivate` combined with `only printing` or so is accordingly bad. So, maybe `Toggle Notation Off "_ + _" (only parsing)`, `Toggle Notation On "_ + _" : scope`.
2. For deactivating all "_ + _" at the same time, maybe `Toggle Notation Off "_ + _" : all`, if really needed
3. No opinion on `#[ enable ] Notation "_ + _" : nat_scope`

Comments and suggestions needed. I'm leaning otherwise towards option 1.

proux01 commented:
> 1. `Deactivate` combined with `only` or so is accordingly bad. So, maybe `Toggle Notation Off "_ + _" (only parsing)`, `Toggle Notation On "_ + _" : scope`.

Does that deactivate parsing or keep parsing?

What about
```
Toggle Notation "_ + _" (only parsing).
Toggle Notation "_ + _" (only printing).
Toggle Notation "_ + _" (off).
Toggle Notation "_ + _" (on).
```
but I still find toggle quite ambiguous (maybe `Set Notation`?).

cpitclaudel commented:
This looks like a wonderful feature.  Most often, though, what I want to do is deactivate a whole scope worth of notations.  `Close Scope` do that; I wonder if it would be easy to implement.  The API would be very simple : `Deactivate Scope xyz_scope`.

Regarding attributes, it seems odd to use them for this.  Are there other cases when the command takes on a completely different meaning when an attribute is added? (The examples I have in mind are like `#[program]`, which tweaks the behavior of an existing command).

Regarding the API of the per-notation feature: it would be great if each line of the output of `Locate "+"` could be used to deactivate the corresponding notation by pasting it directly after `Locate Notation`.  And with a plain string, `Deactivate "+"` would mean disable everything that is returned by `Locate "+"` (this flexibility is very nice in `unfold`, for example)

Thanks a lot for working on this!

Zimmi48 commented:
> Regarding attributes, it seems odd to use them for this. Are there other cases when the command takes on a completely different meaning when an attribute is added? (The examples I have in mind are like `#[program]`, which tweaks the behavior of an existing command).

No indeed. Current attributes tweak the behavior rather than changing it entirely. Maybe what I proposed is stretching their use too far. It didn't get a lot of support anyway.

> but I still find toggle quite ambiguous (maybe `Set Notation`?).

Note that `Set` + any combination of identifiers following it is already taken for flags and options.

herbelin commented:
The general demand seems to be able to describe sets of notations in a variety of ways. So, it seems preferable to seek for a syntax with filters. So, coming back to `Activate`/`Deactivate` (but it could maybe also overload the role of `Set`/`Unset`), a proposed syntax is:
```
["Activate"|"Deactivate"]
  [|"Printing"|"Parsing"]
  ["Notation"|"Notations"]
  [|qualid|string] [|"for" pattern] [|"(in custom " ident ")"] [|":" scope].
```
with the following meaning:
- `Parsing`/`Printing`: restrict activation/deactivation to parsing/printing; applies otherwise to both
- each of optional `ident`/`string`, `pattern`, `(in custom entry)`, `: scope` acts as a filter to select one or more notations
  - `qualid` selects abbreviations ending with `qualid`
  - `string` selects notations with the corresponding syntax, if `string` has no space, it refers to all notations containing the string (like `Locate`)
  - `pattern` selects abbreviations/notations matching the pattern at its head
  - `in custom entry` selects notations in such custom entry
  - `: scope` selects all notations in scope
 - if `Notation` is given, the filter is expected to give exactly one result

Examples:
- `Deactivate Notations "_ + _".` -> deactivate all notations
- `Deactivate Notations for Nat.add.` -> deactivate all notations with head constant `Nat.add`
- `Deactivate Notations "+" for Nat.add.` -> deactivate all notations with head constant `Nat.add`
- `Deactivate Notations : nat_scope` -> deactivate all notations in scope `nat_scope`

Additionally, when all notations associated to a parsing rule have been deactivated, the grammar rule is removed.

For scopes, maybe `in nat_scope` would be clearer than `: nat_scope`? We could also imagine a syntax reminiscent of `Search` (e.g. with `scope:nat_scope`, `pattern:Nat.add`, `head:Nat.add`, ...)?

I'm unsure to be able to find the time to implement all the variety of filters shortly, but the syntax can still be anticipated and bound to the already existing activation/deactivation mechanism

SkySkimmer commented:
If the doc_gram can be fixed this may be ready?

herbelin commented:
From memory, the discussion on the concrete syntax did not really converge, but otherwise, this was usable.

Zimmi48 commented:
At some point, you're the developer of this new feature, so it's your call to decide on a syntax, unless someone voices clear opposition to the chosen solution. I don't think we can expect a natural convergence for any syntax discussion.

That being said, I will try to summarize my current position as shortly as possible:

- I personally have some strong opposition about making `Set` / `Unset` become more polysemous than it is already today.

- I suggested something based on the new attribute syntax, but noticing that the `(only parsing)` and co mechanism is already a preexisting attribute syntax (specific to notations), my suggestion could be amended to just use this syntax (add new `(off)` and `(on)` attributes to toggle off / toggle on an existing notation and be done with it). This is similar to a proposal by @proux01 above, but without the need for the `Toggle` prefix.

herbelin commented:
Trying to combine the different ideas and in particular to replace @proux01's `Toggle` with `on`/`off` flags, as suggested by @Zimmi48, and also, to get even more closer to the syntax of `Notation`, I propose to implement tomorrow the following variant of my previous proposal (`on`/`off` mandatory and otherwise at least one other component among `qualid`, `string`, `pattern`, `custom ident`, `scope`).
```
"Tune" ["Notation"|"Notations"] [|qualid|string] [|":=" pattern]
   "(" [|in custom " ident] [|"parsing"|"printing"] ["on"|"off"] ")"
   [|":" scope].
```
with same semantics as before (mutatis mutandis).

If the general idea is ok, possible synonyms of `Tune` would be `Setup` (to avoir `Set`!), `Alter`, `Adjust`, `Revise`, `Tweak`, `Amend`, ..

jfehrle commented:
Typo: I think `[|in custom " ident]` should be `[| "in" "custom" ident]`

I suggest allowing only "Notation" and omitting "Notations", which might permit the parser to distinguish "Set Notation" as a special case distinct from "Set".  It would break the pattern for "Set" but I think users would quickly catch on.

Distinguishing singular or plural won't affect user understanding, and we don't have to do odd things like this, nor include both versions in the command index.  WDYT?

![image](https://user-images.githubusercontent.com/1253341/141351723-ebb4670f-2aa9-49b4-8f1d-f6e2559c4e94.png)

If "Set" won't work, how about "Transmogrify"? :-)

![image](https://user-images.githubusercontent.com/1253341/141353156-bcbb7ccc-7508-46ef-a2c2-aa76a52d88cc.png)

herbelin commented:
> Typo: I think `[|in custom " ident]` should be `[| "in" "custom" ident]`

Yes. Edited.

> which might permit the parser to distinguish "Set Notation" as a special case distinct from "Set". It would break the pattern for "Set" but I think users would quickly catch on.

I would find it quite natural too. There were some objections though.

> I suggest allowing only "Notation" and omitting "Notations"

> Distinguishing singular or plural won't affect user understanding, and we don't have to do odd things like this, nor include both versions in the command index. WDYT?

The idea was to distinguish between denoting a single notation or possibly denoting a group of notations.

But we can imagine as alternative to have a modifier "all" to explicitly tell to catch all notations matching the criterion.

I still feel that one should be able to give an incomplete pattern expecting that it refers to exactly one notation.

> If "Set" won't work, how about "Transmogrify"? :-)

Why not! The word is amusing.

Zimmi48 commented:
My impression is that `Toggle` was dismissed too fast and none of the new suggestions seem to be any better. `Toggle` in combination with a mandatory `on` / `off` attribute does not sound ambiguous to me.

Anyway, my own suggestion was to avoid putting a verb first, but to keep `Notation` as the first word. I've recently read something that matches my intuition pretty well, so I will quote it instead of rephrasing it poorly:

>Commands should follow a noun-verb dialogue. Although noun-verb formatting seems backwards from a speaking perspective (i.e. nix store copy vs. nix copy store) it allows us to organize commands the same way users think about completing an action (the group first, then the command).

Source: https://nixos.org/manual/nix/unstable/contributing/cli-guideline.html#naming-the-commands

When we look at our command index, we see many cases where putting the verb first have created issues. To start with, we have the `Add` / `Remove` commands for tables that conflict with the `Add X` commands and restrict how we can name tables. Similarly, `Print` is limited in its ability to print objects with some reasonable names because of all the `Print X` commands that exist. We also have not been able to properly turn `Export` into a legacy attribute that could be used (without the new `#[ export ]` syntax) in front of many commands simply because of the conflict with the `Export` command.

Anyway, it seems that it is the path that has been chosen (to match natural language more closely) and that it is too late to divert from it now (except in a new major version of the language, but that's probably not worth it).

proux01 commented:
`Tune`, `Transmogrify` or even `Toggle` look good to me. As a side question, would nothing be a solution?
```
"Notation" [|qualid|string] [|":=" pattern]
   "(" [|"in" "custom" ident] [|"parsing"|"printing"] ["on"|"off"] ")"
   [|":" scope].
```
that would avoid introducing a new keyword

gmalecha commented:
To support activating and deactivating many notations at once, would it be possible to have collections of notations? Perhaps there is some overlap with notation scopes here, but since arguments can bind notation scopes, this requires more than simply `Close Scope ...`.

herbelin commented:
An update on this PR:
- I implemented a "slection engine" for notations allowing parsing/printing activation/deactivation based on scope, custom entry, name or syntax, interpretation (as discussed above)
- I did not try to merge the syntax with the one of `Notation` (fearing it would be complex to analyze). The current syntax is:
  ```
  "Toggle" "Notation" [|qualid|string] [|":=" pattern]
     "(" [|"in" "custom" ident] [|"parsing"|"printing"] ["on"|"off"] ")"
     [|":" scope|":" "no" "scope"].
  ```
  The pattern is currently interpreted as a filter of the actual intepretation and the notation string is interpreted like in `Locate`.
- The parsing rule is currently not removed (even if all notations using the corresponding parsing rule are deactivated).

Here are examples:
```
Toggle Notation := ex2 (off) : type_scope.
(*
The following notations have been deactivated:
Notation "'exists2' x : A , p & q" := (ex2 (fun x => p) (fun x => q))
Notation "'exists2' x , p & q" := (ex2 (fun x => p) (fun x => q))
Notation "'exists2' ' x : A , p & q" := (ex2 (fun x => p) (fun x => q))
Notation "'exists2' ' x , p & q" := (ex2 (fun x => p) (fun x => q))
*)
```
```
Toggle Notation "<=" (off).
(*
The following notations have been deactivated:
Notation "x <= y <= z" := (and (le x y) (le y z))
Notation "x <= y < z" := (and (le x y) (lt y z))
Notation "n <= m" := (le n m)
Notation "x < y <= z" := (and (lt x y) (le y z))
*)
```

jfehrle commented:
Hard to pick a good name for this.  "Toggle" suggests something that alternates between two states like an on/off button in the physical world.  How about "Notation Set" instead?  I suppose this new command only applies to things defined with `Notation` and doesn't apply to `Tactic Notation`, `Number Notation`, `String Notation`, `Ltac2 Notation`, etc.

proux01 commented:
>     * I did not try to merge the syntax with the one of `Notation` (fearing it would be complex to analyze).

That seems a reasonable concern. However I agree with @jfehrle that the "Toggle" word is potentially ambiguous. Maybe just dropping it would be better? The absence of either the string (as in your first example) or the `:=` part (as in your second example) makes it unambiguous anyway.

jfehrle commented:
IIRC the parser has special case code that looks at all the idents at the beginning of a command together (i.e., no optional elements), so if you make "Notation Set" a separate command, it's clever enough to handle correctly.  It may matter which form is first in the mlg.  An easy experiment.  If that's correct, you could also make it "Set Notation ..." (not following the convention for `Set`, though).

herbelin commented:
`Set Notation` and `Unset Notation` would be ok for the parser (there is e.g. the ad hoc `Set Firstorder Solver @tactic`). It is more that it was not "fully consensual". (Personally, I find it rather intuitive.)

proux01 commented:
@herbelin my understanding is that @jfehrle only offers to change the current "Toggle Notation" to "Notation Set" (or "Set Notation"), nothing about (re)introducing some "Unset Notation".

It seems to me that just changing "Toggle Notation" to "Notation Set" (or less preferred "Set Notation") would be enough to reach a reasonable consensus and move towards merging this long awaited PR.

Alizter commented:
What about `Switch`?

SkySkimmer commented on interp/notationextern.ml:
```diff
@@ -0,0 +1,198 @@
+let init () =
+  notations_key_table := KeyMap.empty
+
+let _ =
```
```suggestion
let () =
```

SkySkimmer commented:
>Note: this seems to reopen #12777 (according to the test-suite) but I don't understand why.

Printing of abbreviations (`Prettyp.gallina_print_abbreviation`) uses `without_specific_symbols` to avoid printing as itself. `without_specific_symbols` works through `is_inactive_rule` which you removed.

You need to either implement `without_specific_symbols` with the new mechanism (and remove all the IRuleSet/inactive_notations_table code which is now unused) or put is_inactive_rule back

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,27 @@ Here are examples:
+      | only parsing
+      | only printing
+
+   Control the activation of notation :token:`string` or
+   :ref:`abbreviation <Abbreviations>` :token:`qualid` or all
+   notations for :token:`one_term`. If given, only notations in scope
+   :token:`ident` are affected (or lonely notations for ``no scope``).
+   The :token:`notation_set_flag` enables to selectively disable or
+   enable parsing or printing.
```
```suggestion
   Enables or disables previously defined notations.
   
   :n:`@string`
     The notation string.
     
   :n:`@qualid`
     The qualified name of a previously defined notation or :ref:`abbreviation <Abbreviations>`.
     (explain whether the local/export mechanism applies to notations)

    If given, only notations in scope
   :token:`ident` are affected (or lonely notations for ``no scope``).
   
   :n:`only parsing`
     The notation is enabled or disabled only for parsing.
     
   :n:`only printing`
     The notation is enabled or disabled only for printing.
```

I think this needs a good bit of work to be understandable to users who may be learning or not an expert.  What you have is perhaps reasonable for someone who knows the feature well.

Some details to address:
- First sentence is too dense.
- "control the activation" is vague.  My guess is that the purpose of the command is to dynamically enable and disable previously defined notations.  If correct, please say that.  "Set" and "Unset" suggest creating or removing a notation, perhaps they should be "Enable" and "Disable"
- If correct, say that `string` is just a notation string and `qualid` is the name of a notation **that was previously defined** with `Notation`.  If it could be from `Reserved Notation`, mention that, too.
- I don't understand "all notations for :token:`one_term`" at all.  Probably deserves an example, perhaps in `.. coqtop::` form
- Need to explain `in custom @ident` and `in constr`

I'll take another look after you've revised this.

proux01 replied:
>     * First sentence is too dense.

Indeed, thanks for the suggestion
 
>     * "control the activation" is vague.  My guess is that the purpose of the command is to dynamically enable and disable previously defined notations.  If correct, please say that.  "Set" and "Unset" suggest creating or removing a notation, perhaps they should be "Enable" and "Disable"

Right, changed to `Enable` and `Disable`
 
>     * If correct, say that `string` is just a notation string and `qualid` is the name of a notation **that was previously defined** with `Notation`.  If it could be from `Reserved Notation`, mention that, too.

done
 
>     * I don't understand "all notations for :token:`one_term`" at all.  Probably deserves an example, perhaps in `.. coqtop::` form

Maybe, let me know if you still think this deserves small examples. I could easily add a few.

>     * Need to explain `in custom @ident` and `in constr`

done
 
>     * Should "lonely notations" be listed in the glossary?

Indeed, done

jfehrle reviewed (commented):
The doc needs some attention.

Note that lint fails because some of your earlier commits don't compile.

jfehrle commented:
PS Does the title "Deactivation of notations: providing vernac syntax, vo registration and support for deactivation of abbreviations" match the PR?  I didn't notice anything about vo files (though I didn't read the code).  Maybe something like this is clearer: "Activation and deactivation of notations and abbreviations"

PPS A changelog will eventually be needed.

herbelin commented:
@proux01:
> @herbelin I rebased this and changed `Toggle Notation ... (on)`/`Toggle Notation ... (off)` to `Notation Set ...`/`Notation Unset ...` and started a first documentation. Hope this will enable to reach a consensu on the syntax and move towards finally merging the feature.

Thanks!

I'll assume that you continue to take over (but we can also interact if you wish).

@jfehrle:
> PS Does the title "Deactivation of notations: providing vernac syntax, vo registration and support for deactivation of abbreviations" match the PR? I didn't notice anything about vo files (though I didn't read the code). Maybe something like this is clearer: "Activation and deactivation of notations and abbreviations"

The vo registration is introduced by the `let inNotationActivation ...` in `metasyntax.ml`. The original motivation for the title was to tell what #703 did not do. The PR evolved since then, so a shortest title would be ok too. Also, the ability to select several notations at once could be mentioned.

proux01 commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,27 @@ Here are examples:
+~~~~~~~~~~~~~~~~~~~~~~
+
+.. cmd:: Notation Set {? {| @string | @qualid } } {? := @one_term } {? ( {+, @notation_set_flag } ) } {? {| : no scope | : @ident } }
+         Notation Unset {? {| @string | @qualid } } {? := @one_term } {? ( {+, @notation_set_flag } ) } {? {| : no scope | : @ident } }
```
> Since the right hand sides are the same, how about combining them

Right, done

> I see everything on the right hand side is optional. Is that syntactically correct? Are there semantic checks that something that some of them are present? (What, if anything would Notation Set. do?

Added a sentence and documented the error is nothing is present.

> Also, would you rename @ident to @scope, @scope_name or @scope_key to correspond to its semantics?

I'd love to but I was unable to find the correct `REPLACE` statement to add in `common_edit.mlg`

proux01 commented:
Thanks all for the reviews, they are now taken into account and CI is green.

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. cmd:: {| Enable | Disable } Notation {? {| @string | @qualid } } {? := @one_term } {? ( {+, @toggle_notation_flag } ) } {? {| : no scope | : @ident } }
+   :name: Disable Notation
```
```suggestion
   :name: Enable Notation; Disable Notation
```

So both forms appear in the index

jfehrle reviewed (commented):
Several small suggestions

proux01 commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+   Enables or disables notations previously defined with
+   :cmd:`Notation` or :cmd:`Notation (abbreviation)`. At least one of
+   :token:`string`, :token:`qualid`, :token:`one_term` or scope must be
+   provided. If more than one is provided, the notations enabled or
```
Here is an attempt. The combinatorics means 7 possibilities. This feels a bit verbose to me but no strong opinion, if you think it's better, I'll trust you.

jfehrle replied:
Yeah, too verbose, not what I meant.  Listing all the possibilities forces the reader to compare them to find how they differ.  I just mean to remove the `{? ... }` in `{? ( {+, @toggle_notation_flag } ) }` if that's correct (apparently not).  I guess your original sentence is correct.  What error does `Enable Notation.` give, if any?

proux01 replied:
> apparently not

Indeed, this part is (unfortunately) not mandatory (omiting it is even likely the most common case)

> What error does Enable Notation. give, if any?

`.. exn:: No notation provided. ` as documented below

proux01 commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+      :ref:`Abbreviation <Abbreviations>` to enable or disable.
+
+   :n:`{? := @one_term }`
+      Enable or disable notations involving :token:`one_term`.
```
Indeed, is `containing` better ?

jfehrle replied:
How about ```matching :token:`one_term` ```?  IIUC that's fundamentally what it's about.  If correct, maybe say matching is based on the AST after expanding any notations in `one_term`.  (`one_term` can be written using notations or not as the user prefers, right?  (perhaps worth mentioning?)

proux01 commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+      Enable or disable notations in given :ref:`custom entry
+      <custom-entries>`.
+
+   :n:`in constr`
+      Enable or disable notations in ``constr``. See
+      :ref:`custom entries <custom-entries>`.
```
Well, constr here is a particular case of custom entry, hence the explicit reference to custom entries where this is explained (after a few paragraphs, maybe this would require a more precise reference).

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+   :n:`{? := @one_term }`
+      Enable or disable notations involving :token:`one_term`.
+
+   :n:`{? {| : no scope | : @ident } }`
```
Helloooo out there, you skipped this one again.  Please do it or say why not.

proux01 replied:
Sorry, I was sure I answered but obviously not. As I said above https://github.com/coq/coq/pull/12324#discussion_r1009212477, I'm unfortunately unable to write the appropriate thing in `common_edit.mlg`. But you're perfectly right this should be `scope_name`.

Here is what I tried
```
| REPLACE [ "Enable" | "Disable" ] "Notation" OPT [ string | qualid ] OPT ( ":=" one_term ) OPT ( "(" LIST1 enable_notation_flag SEP "," ")" ) OPT [ ":" ident | ":" "no" "scope" ]
| WITH    [ "Enable" | "Disable" ] "Notation" OPT [ string | qualid ] OPT ( ":=" one_term ) OPT ( "(" LIST1 enable_notation_flag SEP "," ")" ) OPT [ ":" scope_name | ":" "no" "scope" ]
```
(copied from `orderedGrammar`)
but I get the following error
```
make doc_gram_rsts
Error: Undefined nonterminal `one_term` in edit: [ "Enable" | "Disable" ] "Notation" OPT [ string | qualid ] OPT ( ":=" one_term ) OPT ( "(" LIST1 enable_notation_flag SEP "," ")" ) OPT\
 [ ":" scope_name | ":" "no" "scope" ]
Error: Can't find '[ "Enable" | "Disable" ] "Notation" OPT [ string | qualid ] OPT ( ":=" one_term ) OPT ( "(" LIST1 enable_notation_flag SEP "," ")" ) OPT [ ":" ident | ":" "no" "scope\
" ]' in edit for 'syntax'
2 error(s)
```
Any help here would be welcome.

jfehrle replied:
Easily solved.  `constr` is renamed to `one_term` towards the end of the file, after your edit.  If you temporarily insert `| PRINT` at the point of your failing edit, `doc_grammar` will print the current definition of the containing nonterminal (perhaps `simple_tactic`).  Probably just use `constr` here instead of `one_term` in both lines.

proux01 replied:
Thanks a lot, that worked like a charm!

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+
+   Enables or disables notations previously defined with
+   :cmd:`Notation` or :cmd:`Notation (abbreviation)`.
+   This command has no effect on parser modifications made with
+   :cmd:`Reserved Notation`.
+   When multiple
+   informations are provided, the notations enabled or
+   disabled are the ones meeting the conjunction of all constraints.
```
```suggestion
   It has no effect on notations reserved with :cmd:`Reserved Notation`.
   When multiple clauses are provided, the notations enabled or
   disabled must satisfy all of their constraints.
```

Best not to introduce a second phrase "parser modifications"
English doesn't use "informations".  "information" is an uncountable noun.  If you want to emphasize that there are multiple items, pick a different noun.

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+
+   :n:`all`
+      Enable or disable all notations meeting the given constraints,
+      even if multiple ones. When not provided, there must be a single
+      notation meeting the constraints.
```
```suggestion
      even if there are multiple ones. Otherwise there must be a single
      notation meeting the constraints.
```

What error do you get if there are multiple matches?

proux01 replied:
```
.. exn:: More than one interpretation bound to this notation, confirm with the "all" modifier.
```
as documented below

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+      <custom-entries>`.
+
+   :n:`in constr`
+      Enable or disable notations in ``constr``. See
```
```suggestion
      Enable or disable notations in the custom entry for :n:`constr`. See
```

Yes?

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+      Enable or disable notations in ``constr``. See
+      :ref:`custom entries <custom-entries>`.
+
+   :n:`{? {| : no scope | : @ident } }`
+      If given, only notations in scope :token:`ident` are affected (or
```
```suggestion
   :n:`{| : no scope | : @ident }`
      Only notations in scope :token:`ident` are affected (or
```

If `ident` is the more common case, I'd put that before `no scope` everywhere (mlgs, doc).  Your explanation here covers `ident` first.  And shouldn't `ident` be something like `scope` (scope_name, scope_key)?

proux01 replied:
> If `ident` is the more common case, I'd put that before `no scope` everywhere (mlgs, doc). Your explanation here covers `ident` first.

Right, done

> And shouldn't `ident` be something like `scope` (scope_name, scope_key)?

It should indeed be `scope_name`, see above comment https://github.com/coq/coq/pull/12324#discussion_r1011379013

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+      Cannot enable or disable for parsing a notation that was
+      originally defined as only printing.
+
+   .. exn:: Found no matching notation to enable.
+
+      No previously defined notation satisfies the given constraints.
+
+   .. exn:: Found no matching notation to disable.
+
+      No previously defined notation satisfies the given constraints.
```
These could be one message.  From the command syntax, it will be obvious whether it's enabling or disabling.

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,109 @@ Here are examples:
+
+   .. exn:: More than one interpretation bound to this notation, confirm with the "all" modifier.
+
+      Use the ``all`` option to allow enabling or disabling multiple
```
```suggestion
      Use the :n:`all` to allow enabling or disabling multiple
```

jfehrle reviewed (commented):
Getting better, a few more details

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,106 @@ Here are examples:
+   :cmd:`Notation` or :cmd:`Notation (abbreviation)`.
+   It has no effect on notations reserved with :cmd:`Reserved Notation`.
+   At least one of
+   :token:`string`, :token:`qualid`, :token:`one_term` or scope must be
```
```suggestion
   :token:`string`, :token:`qualid`, :token:`one_term` or :token:`scope_name` must be
```

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,106 @@ Here are examples:
+
+   :n:`all`
+      Enable or disable all notations meeting the given constraints,
+      even if multiple ones. Otherwise, there must be a single
```
```suggestion
      even if there are multiple ones. Otherwise, there must be a single
```

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,106 @@ Here are examples:
+      Enable or disable notations in the custom entry for :n:`constr`.
+      See :ref:`custom entries <custom-entries>`.
+
+   :n:`{? {| : @ident | : no scope } }`
```
```suggestion
   :n:`{| : @ident | : no scope }`
```

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,106 @@ Here are examples:
+      See :ref:`custom entries <custom-entries>`.
+
+   :n:`{? {| : @ident | : no scope } }`
+      If given, only notations in scope :token:`ident` are affected (or
```
```suggestion
      If given, only notations in scope :token:`scope_name` are affected (or
```

jfehrle reviewed (commented):
Converging...
Please add a changelist and update the checkboxes in the PR description.

proux01 commented:
> Converging... Please add a changelist and update the checkboxes in the PR description.

Thanks, done

jfehrle commented on doc/changelog/03-notations/12324-master+nametab-method-remove.rst:
```diff
@@ -0,0 +1,5 @@
+- **Added:**
+  command :cmd:`Enable Notation`
```
```suggestion
  :cmd:`Enable Notation` command
```

jfehrle commented on doc/changelog/03-notations/12324-master+nametab-method-remove.rst:
```diff
@@ -0,0 +1,5 @@
+  command :cmd:`Enable Notation`
+  to enable or disable previously defined notations
+  (`#12324 <https://github.com/coq/coq/pull/12324>`_,
+  by Hugo Herbelin).
```
Only Hugo, not Pierre, too?

herbelin replied:
Yes, Pierre too. I pushed refined credits.

proux01 replied:
I feel I only fixed a few final points and wrote the doc, but why not.
I also credited Jim's review which, as always, greatly improved the quality of the documentation.

jfehrle replied:
Now you need someone to review the code...

jfehrle commented on doc/sphinx/user-extensions/syntax-extensions.rst:
```diff
@@ -441,6 +441,106 @@ Here are examples:
+
+   :n:`{| : @scope_name | : no scope }`
+      If given, only notations in scope :token:`scope_name` are affected (or
+      :term:`lonely notations <lonely notation>` for ``no scope``).
```
```suggestion
      :term:`lonely notations <lonely notation>` for :n:`no scope`).
```

jfehrle reviewed (approved):
LGTM.  A few last details

proux01 commented:
@coqbot run full ci now

proux01 commented:
@coqbot run full ci now

proux01 commented:
CI "green", I'll merge on Thursday if there are no more comments.

jfehrle commented:
As a co-author, shouldn't this be assigned to someone else other than you, Pierre?  (And my review applies only to the documentation.)

proux01 commented:
> As a co-author, shouldn't this be assigned to someone else other than you, Pierre? (And my review applies only to the documentation.)

Honestyl, I still consider myself as a reviewer here. I didn't really wrote any code, only slightly edited Hugo's code. Sure, ideally an additional review would be great but given our human ressources, it's probably best not to wait for it.

Anyway @ejgallego you're still marked as assignee here (since last year), let us know if you intend to have a closer look, otherwise I'll merge.

jfehrle commented:
That's fine with me.

proux01 commented:
@coqbot merge now

proux01 commented:
@coqbot merge now

ejgallego commented:
> Anyway @ejgallego you're still marked as assignee here (since last year), let us know if you intend to have a closer look, otherwise I'll merge.

Sorry I don't have a very fast turn around time lately; indeed I was planning to take care of this PR; I'll have a look and provide some comments ASAP.

proux01 commented:
Well, it's merged now so unless you see something that you think really needs fixing, you can save your time.

ejgallego commented:
> Well, it's merged now so unless you see something that you think really needs fixing, you can save your time.

I need to have a look, but indeed I'm convinced that the vernac-based interface to interact with notations is IMO the wrong one, as it doesn't scale except for very small cases. Interactively, the users want to understand what's going on "on the fly", and not in a static way, which clutters the document.

Isabellle got this right, and hopefully once we advance with the coq-layout-engine we can do like that, so the API here should be designed in a way that is compatible.

herbelin commented:
> > Well, it's merged now so unless you see something that you think really needs fixing, you can save your time.
> 
> I need to have a look, but indeed I'm convinced that the vernac-based interface to interact with notations is IMO the wrong one, as it doesn't scale except for very small cases. Interactively, the users want to understand what's going on "on the fly", and not in a static way, which clutters the document.

Do you mean that activation/inactivation (and printing) should be managed only by the interfaces?

If not, the question of a vernac-accessible API to activation/deactivation does not seem contradictory with a direct access from the UI to the activation/deactivation API? In particular, I agree that from a UI, we don't want to issue a vernac command but rather use the mouse, menus, or interactive activation/deactivation widgets.

herbelin commented:
I'm sorry, but I made a mistake when saying that it fixed #4803. The PR does not remove rules from the parser.

To do so, one would typically need to check after deactivation if some interpretation of a parsing rule still exists, and if not, to return to `metasyntax.ml` the list of orphan rules, so that `metasyntax.ml` can recover the rule (`recover_notation_syntax`) and ask `pcoq.ml` to register a deletion of the rule (one need for that to add to `pcoq.ml` the ability to delete, calling `Grammar.safe_delete_rule`, in a way synchronous with backtracking).

Zimmi48 commented:
Should the documentation clarify that this does not remove the parsing rules (so `Notation` + `Disable Notation` has a behavior similar to `Register Notation`)?

ejgallego commented on library/nametab.ml:
```diff
@@ -212,6 +210,36 @@ let push visibility uname o tab =
+      in
+      mktree this tree.map
+
+let remove uname tab =
```
@herbelin can we document what this function does?

herbelin replied:
I added a comment of this function in #16945.

maximedenes commented:
> I need to have a look, but indeed I'm convinced that the vernac-based interface to interact with notations is IMO the wrong one, as it doesn't scale except for very small cases. Interactively, the users want to understand what's going on "on the fly", and not in a static way, which clutters the document.

I concur. Fine grained control notations is obviously a very nice feature, but I find it strange to add even more complexity to the vernacular jungle. Out of curiosity, were there many use cases needing to have this notation control persist in a final document?

proux01 commented:
It seems the main use case we had in the discussion above was @gares needing to selectively disable notations from MathComp for teaching purpose. So it seems now possible to provide a file `MC_course.v` with appropriate commands that works with a distributed version of the library (without having to rebuild it). It's true that for this use case a command that would only live in the current session rather than persisting would probably have been enough, but this means the student would have seen all the ugly `Deactivate Notation` at the beginning of their exercise file, maybe not ideal.

maximedenes commented:
This was merged with TODOs and stubs in the code. What they imply is not clear to me (and not documented AFAICT). Is the plan to address them in the very short term?

herbelin commented:
From @maximedenes:
> I concur. Fine grained control notations is obviously a very nice feature, but I find it strange to add even more complexity to the vernacular jungle. Out of curiosity, were there many use cases needing to have this notation control persist in a final document?

From @proux01:
> It seems the main use case we had in the discussion above was @gares needing to selectively disable notations from MathComp for teaching purpose. 

I also think that the main use is in interactive mode (teaching or more generally proof development). My dream is that this control is eventually driven by GUI (control-click on a notation, global selection panel, etc.), but it has to be implemented somewhere anyway. So, where to implement it if not in the code of Coq itself (which has all the information to do the filtering job)?

proux01 commented:
> I also think that the main use is in interactive mode

Note that the use case above is not necessarily purely interactive. One may want to provide a file with a few `Disable Notation` commands, even for teaching purpose (where asking students to click on a bunch of notations might be extremely inconvenient). Sure a GUI would ultimately be nice, but it seems GUI and vernac command have complementary use cases.

maximedenes commented:
> I also think that the main use is in interactive mode (teaching or more generally proof development). My dream is that this control is eventually driven by GUI (control-click on a notation, global selection panel, etc.)

+1

> but whatever the way the command is driven, the code for the command has to be implemented somewhere anyway. So, where to implement it if not in the code of Coq itself (which has all the information to do the filtering job)?

Yes, the code needs to be in Coq's notation engine. What I find much more debatable is to have it wired in vernac. In particular, I imagine there's underneath a libobject which depends on other libobjects (like the one initially definining the notation). This usually comes with many corner cases.

> In particular, this seems similar to me to the way `Set Printing All` is driven by user interfaces without needing it to be a line in a document.

I wouldn't take what is done today by GUIs on this particular point as a model :)

maximedenes commented:
> Note that the use case above is not necessarily purely interactive. One may want to provide a file with a few `Disable Notation` commands, even for teaching purpose (where asking students to click on a bunch of notations might be extremely inconvenient). Sure a GUI would ultimately be nice, but it seems GUI and vernac command have complementary use cases.

The GUI could come with profiles or scripting capabilities, rather than cluttering the vernac language to allow for monkey patching of libraries (not the simplest thing to explain to students, if you ask me).

herbelin commented:
The fact is that the vernal level is not only the language for document, it is also a (non-graphical) UI. So, maybe, what you are asking for is to split the vernac in two parts (which was actually the idea of isolating a Gallina specification language beside the language used for driving the development environment).

SkySkimmer commented:
>Yes, the code needs to be in Coq's notation engine. What I find much more debatable is to have it wired in vernac. In particular, I imagine there's underneath a libobject which depends on other libobjects

It's possible to have a command without a libobject though
It just means effects won't persist after module/section/file end

jfehrle commented:
> The GUI could come with profiles or scripting capabilities, rather than cluttering the vernac language to allow for monkey patching of libraries (not the simplest thing to explain to students, if you ask me).

Better to implement that once in Coq itself and make it available in each GUI.  Also better to avoid GUI-specific scripting.  

> One may want to provide a file with a few `Disable Notation` commands, even for teaching purpose (where asking students to click on a bunch of notations might be extremely inconvenient).

There would be a question of which panels support clicking on a notation.  If you want to support it in both script panels and in the context panel, their contents are quite different and would require separate code to support each.

herbelin commented:
I would say that the context panel is the primary use. I don't know if it is already part of the roadmap of someone or if we can just start to experiment.
